capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+        // array), but it's better than using unsafe.  This
+        // implementation is not intended to be high performance.
+        let mut message_schedule: [u32; 64] = [0; 64];
+        for i in 0..16 {
+            let val: u32 = (buffer[i * 4 + 0] as u32) << 24
+                | (buffer[i * 4 + 1] as u32) << 16
+                | (buffer[i * 4 + 2] as u32) << 8
+                | (buffer[i * 4 + 3] as u32);
+            message_schedule[i] = val;
+        }
+        self.perform_sha(&mut message_schedule);
+    }
+
+    fn compute_block(&self, data: &mut [u8; 64]) {
+        self.compute_buffer(data);
+    }
+
+    fn perform_sha(&self, message_schedule: &mut [u32; 64]) {
+        // Message schedule
+        for i in 16..64 {
+            let mut s0 = self.right_rotate(message_schedule[i - 15], 7);
+            s0 ^= self.right_rotate(message_schedule[i - 15], 18);
+            s0 ^= message_schedule[i - 15] >> 3;
+            let mut s1 = self.right_rotate(message_schedule[i - 2], 17);
+            s1 ^= self.right_rotate(message_schedule[i - 2], 19);
+            s1 ^= message_schedule[i - 2] >> 10;
+            message_schedule[i] = message_schedule[i - 16] + s0 + message_schedule[i - 7] + s1;
+        }
+
+        // Compression
+        let mut hashes = self.hash_values.get();
+        for i in 0..64 {
+            let s1 = self.right_rotate(hashes[4], 6)
+                ^ self.right_rotate(hashes[4], 11)
+                ^ self.right_rotate(hashes[4], 25);
+            let ch = (hashes[4] & hashes[5]) ^ ((!hashes[4]) & hashes[6]);
+            let constant = ROUND_CONSTANTS[i];
+            let temp1 = hashes[7] + s1 + ch + constant + message_schedule[i];
+            let s0 = self.right_rotate(hashes[0], 2)
+                ^ self.right_rotate(hashes[0], 13)
+                ^ self.right_rotate(hashes[0], 22);
+            let maj = (hashes[0] & hashes[1]) ^ (hashes[0] & hashes[2]) ^ (hashes[1] & hashes[2]);
+            let temp2 = s0 + maj;
+
+            hashes[7] = hashes[6];
+            hashes[6] = hashes[5];
+            hashes[5] = hashes[4];
+            hashes[4] = hashes[3].wrapping_add(temp1);
+            hashes[3] = hashes[2];
+            hashes[2] = hashes[1];
+            hashes[1] = hashes[0];
+            hashes[0] = temp1.wrapping_add(temp2);
+        }
+
+        let mut new_hashes = self.hash_values.get();
+        for i in 0..8 {
+            new_hashes[i] = new_hashes[i].wrapping_add(hashes[i]);
+        }
+        self.hash_values.set(new_hashes);
+    }
+}
+
+impl<'a> DigestData<'a, 32> for Sha256Software<'a> {
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])> {
+        if self.busy() {
+            Err((ErrorCode::BUSY, data.take()))
+        } else {
+            self.state.set(State::Data);
+            if self.handle.is_none() {
+                Err((ErrorCode::FAIL, data.take()))
```

Comment:
Won't this leave the state stuck in `Data`?

Reply:
No -- `call` moves it to `Idle`.